cmake/detect_gpu.cmake

```diff
@@ -0,0 +1,215 @@
+        "  return 0;\n"
+        "}\n")
+
+    execute_process(COMMAND "${CMAKE_CUDA_COMPILER}" "--run" "${__cufile}"
+                    WORKING_DIRECTORY "${PROJECT_BINARY_DIR}/CMakeFiles/"
+                    RESULT_VARIABLE __nvcc_res OUTPUT_VARIABLE __nvcc_out
+                    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    if(__nvcc_res EQUAL 0)
+        set(CUDA_gpu_detect_output ${__nvcc_out} CACHE INTERNAL
+            "Returned GPU architetures from ginkgo_detect_gpus tool" FORCE)
+    endif()
+
+    if(NOT CUDA_gpu_detect_output)
+        message(STATUS "Automatic GPU detection failed.
+                       Building for all known architectures.")
+        set(${out_variable} ${ginkgo_known_gpu_archs} PARENT_SCOPE)
+    else()
+        set(${out_variable} ${CUDA_gpu_detect_output} PARENT_SCOPE)
+    endif()
+endfunction()
+
+
+################################################################################
+# A function for checking whether it is available
+# Usage:
+#     check_available(arch_var)
+function(check_available arch_var)
+    if(arch_var MATCHES "([0-9]+)")
+        if(NOT(arch_var IN_LIST cuda_all_arch_list))
+            message(FATAL_ERROR "arch ${arch_var} is not invalid")
+        elseif(arch_var IN_LIST ginkgo_known_gpu_archs)
+            if(arch_var IN_LIST ginkgo_unsupported_archs)
+                message(FATAL_ERROR "Ginkgo does not support ${arch_var}")
+            endif()
+        else()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    else()
+        if(arch_var IN_LIST ginkgo_known_gpu_archs_name)
+            if(arch_var IN_LIST ginkgo_unsupported_archs_name)
+                message(FATAL_ERROR
+                    "Ginkgo does not support ${arch_var}")
+            endif()
+        else ()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    endif()
+endfunction()
+
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_OPTION
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    if(CMAKE_CUDA_COMPILER_VERSION MATCHES "([0-9]+).([0-9]+).(.*)")
+        set(cuda_version ${CMAKE_MATCH_1}${CMAKE_MATCH_2})
+    endif()
+    set(ginkgo_known_gpu_archs ${ginkgo_known_gpu_archs_cuda_${cuda_version}})
+    set(ginkgo_known_gpu_archs_name
+        ${ginkgo_known_gpu_archs_name_${cuda_version}})
+    set(cuda_arch_bin_Volta ${cuda_arch_bin_Volta_${cuda_version}})
+    set(cuda_all_arch_list "")
+    foreach(__ver ${ginkgo_known_cuda_version})
+        list(APPEND cuda_all_arch_list ${ginkgo_known_gpu_archs_cuda_${__ver}})
+    endforeach()
+    list(REMOVE_DUPLICATES cuda_all_arch_list)
+
+    # set CUDA_ARCH_OPTION strings
+    set(CUDA_ARCH_OPTION "Auto" CACHE STRING
```

Comment:
A minor nit: could we move this variable definition to the rest of the options in the top-level `CMakeListst.txt` file. This way all the options are defined in the same place, so users can quickly see which options we support.

Reply:
move CUDA_ARCH_OPTION to top-level now